File: python/examples/gfa2_info/gfa2_info.py
#!/usr/bin/env python3
"""
Validates a GFA2 file and display content statistics

Usage: ./gfa2_info.py <inputfile> <gfa2spec>

Arguments:
  <inputfile>  Input file in GFA2 format
  <gfa2spec>   Specification file describing the GFA2 format
"""
from textformats import Specification
from docopt import docopt
from gfa2_stats_collector import Gfa2StatsCollector
from gfa2_cross_validator import Gfa2CrossValidator
import sys

def stats_collector_process_gfaline(line, stats_collector):
  stats_collector.lt(line["line_type"])
  if line["line_type"] == "segment":
    stats_collector.seq(line["slen"], line["sequence"] is not None)
  for tn, content in line.get("tags", {}).items():
    stats_collector.tag(tn, content["type"])

def process_gfaline(line, data):
  stats_collector, validator = data
  stats_collector_process_gfaline(line, stats_collector)
  validator.process_gfaline(line)

def main(args):
  gfa2spec = Specification(args["<gfa2spec>"])
  validator = Gfa2CrossValidator()
  stats_collector = Gfa2StatsCollector(validator.ids)
  gfa2spec["line"].decode_file(args["<inputfile>"], process_gfaline,
                               (stats_collector, validator))
  validator.post_validations()
  if validator.n_err > 0:
    sys.stderr.write(f"GFA2 cross-validations failed\n")
    sys.stderr.write(f"Total number of errors: {validator.n_err}\n")
    sys.exit(1)
  else:
    print(stats_collector, end="")
    sys.exit(0)

if __name__ == "__main__":
  args = docopt(__doc__)
  main(args)

File: examples/gfa2_info/gfa2_info.nim
import os, strutils, strformat, json
import textformats
import gfa2_stats_collector, gfa2_cross_validator

const HelpMsg = """
Validates a GFA2 file and display content statistics

Usage: $# <inputfile> <gfa2spec>

Arguments:
  <inputfile>  Input file in GFA2 format
  <gfa2spec>   Specification file describing the GFA2 format
"""

proc stats_collector_process_gfaline(line: JsonNode,
                                     stats_collector: var Gfa2StatsCollector) =
  stats_collector.lt(line["line_type"].getStr())
  if line["line_type"].getStr() == "segment":
    stats_collector.seq(line["slen"].getInt(),
                        line["sequence"].kind == JString)
  if "tags" in line:
    for tn, content in line["tags"]:
      stats_collector.tag(tn, content["type"].getStr())

type Data = ref object
  stats_collector*: Gfa2StatsCollector
  validator*: Gfa2CrossValidator

proc process_gfaline(line: JsonNode, dataptr: pointer) =
  var data = cast[Data](dataptr)
  stats_collector_process_gfaline(line, data.stats_collector)
  data.validator.process_gfaline(line)

proc parse_args(): (string, Specification) =
  if (paramCount() != 2):
    echo(HelpMsg % [getAppFilename()])
    quit(0)
  result[0] = paramStr(1)
  result[1] = specification_from_file(paramStr(2))

when isMainModule:
  #try:
    let
      (input_file, spec) = parse_args()
      ddef = get_definition(spec, "line")
    var
      validator = newGfa2CrossValidator()
      stats_collector = newGfa2StatsCollector(validator.ids)
      data = Data(validator: validator, stats_collector: stats_collector)
    input_file.decode_file(ddef, false, process_gfaline,
                           cast[pointer](data), DplWhole)
    data.validator.post_validations()
    if data.validator.n_err > 0:
      stderr.write("GFA2 cross-validations failed\n")
      stderr.write(&"Total number of errors: {data.validator.n_err}\n")
      quit(1)
    else:
      echo(data.stats_collector)
      quit(0)
    quit(0)
  #except:
  #  echo(getCurrentExceptionMsg())
  #  quit(1)

